
# FlaskForm class definitions for GeneDescription
# This file is automatically generated from the schema by schema/build_from_schema.py

from flask_wtf import FlaskForm
from flask_wtf.file import FileField
from custom_validators import *
from wtforms import StringField, SelectField, DateField, BooleanField, IntegerField, DecimalField, TextAreaField, HiddenField, validators, MultipleFileField
class GeneDescriptionForm(FlaskForm):
    maintainer = StringField('Curator', [validators.Length(max=255)], description="Maintainer of this sequence record")
    lab_address = StringField('Curator address', [validators.Length(max=255)], description="Institution and full address of corresponding author")
    imgt_name = StringField('IUIS Name', [validators.Length(max=255)], description="The name of this sequence as assigned by IUIS")
    sequence_name = StringField('Sequence Name', [validators.Length(max=255), NonEmpty()], description="The canonical name of this sequence as assigned by IARC")
    alt_names = StringField('Alternative names', [validators.Length(max=255)], description="Alternative names for this sequence")
    chromosome = IntegerField('Chromosome', [validators.Optional()], description="chromosome on which the gene is located")
    mapped = BooleanField('Mapped', [], description="Whether the gene has been mapped to a specific location in the chromosome")
    locus = SelectField('Locus', choices=[('IGH', 'IGH'), ('IGK', 'IGK'), ('IGL', 'IGL'), ('TRA', 'TRA'), ('TRB', 'TRB'), ('TRG', 'TRG'), ('TRD', 'TRD')], description="Gene locus")
    sequence_type = SelectField('Sequence Type', choices=[('V', 'V'), ('D', 'D'), ('J', 'J'), ('C', 'C')], description="Sequence type (V, D, J, C)")
    functionality = SelectField('Functionality', choices=[('F', 'F'), ('ORF', 'ORF'), ('P', 'P')], description="Functionality")
    inference_type = SelectField('Inference Type', choices=[('Genomic Only', 'Genomic Only'), ('Rearranged Only', 'Rearranged Only'), ('Genomic and rearranged', 'Genomic and rearranged')], description="Type of inference(s) from which this gene sequence was inferred (Unrearranged and Rearranged, Unrearranged Only, Rearranged Only)")
    affirmation_level = SelectField('Affirmation Level', choices=[('0', '0'), ('1', '1'), ('2', '2'), ('3', '3')], description="Count of independent studies in which this allele as been affirmed by IARC (1,2,3 or more)")
    species_subgroup = StringField('Species subgroup', [validators.Length(max=255)], description="Race, strain or other species subgroup to which this subject belongs")
    species_subgroup_type = SelectField('Subgroup type', choices=[('none', 'none'), ('breed', 'breed'), ('strain', 'strain'), ('inbred', 'inbred'), ('outbred', 'outbred'), ('locational', 'locational')], description="Category of subgroup")
    gene_subgroup = StringField('Gene Subgroup', [validators.Length(max=255)], description="Gene subgroup (family), as (and if) identified for this species and gene")
    subgroup_designation = StringField('Gene Designation', [validators.Length(max=255)], description="Gene designation within this subgroup, if identified")
    allele_designation = StringField('Allele Designation', [validators.Length(max=255)], description="Allele designation, if identified")
    sequence = TextAreaField('Full Sequence', [ValidNucleotideSequence(ambiguous=False, dot=True)], description="nt sequence of the gene. This should cover the full length that is available, including where possible 5' UTR and lead-in for V-gene sequences")
    coding_seq_imgt = TextAreaField('Coding Sequence', [ValidNucleotideSequence(ambiguous=False, gapped=True)], description="nucleotide sequence of the coding region, aligned, in the case of a V-gene, with the IMGT numbering scheme. For a heavy chain C gene, this should be the secretory form.")
    transmembrane_coding_sequence = TextAreaField('TM Coding Sequence', [validators.Length(max=10000)], description="Nucleotide sequence of the coding region of the transmembrane domain of an IGH/IGI C-gene). The sequence of the cytoplasmic form and the sequence of the coding domain of all other C-genes should be placed in coding_sequence.")
    gene_start = IntegerField('Gene start', [validators.Optional()], description="Co-ordinate (in the sequence field) of the first nucleotide in the coding_sequence field")
    gene_end = IntegerField('Gene end', [validators.Optional()], description="Co-ordinate (in the sequence field) of the last gene-coding nucleotide in the coding_sequence field")
    utr_5_prime_start = IntegerField('UTR 5\' Start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of 5 prime UTR (V-genes only)")
    utr_5_prime_end = IntegerField('UTR 5\' End', [validators.Optional()], description="End co-ordinate in the Full Sequence of 5 prime UTR (V-genes only)")
    leader_1_start = IntegerField('L-PART1 Start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of L-PART1 (V-genes only)")
    leader_1_end = IntegerField('L-PART1 End', [validators.Optional()], description="End co-ordinate in the Full Sequence of L-PART1 (V-genes only)")
    leader_2_start = IntegerField('L-PART2 Start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of L-PART2 (V-genes only)")
    leader_2_end = IntegerField('L-PART2 End', [validators.Optional()], description="End co-ordinate in the Full Sequence of L-PART2 (V-genes only)")
    cdr1_start = IntegerField('CDR1 Start', [validators.Optional()], description="CDR1 start co-ordinate in the Full Sequence (V-genes only)")
    cdr1_end = IntegerField('CDR1 End', [validators.Optional()], description="CDR1 end co-ordinate in the Full Sequence (V-genes only)")
    cdr2_start = IntegerField('CDR2 Start', [validators.Optional()], description="CDR2 start co-ordinate in the Full Sequence (V-genes only)")
    cdr2_end = IntegerField('CDR2 End', [validators.Optional()], description="CDR2 end co-ordinate in the Full Sequence (V-genes only)")
    cdr3_start = IntegerField('CDR3 Start', [validators.Optional()], description="CDR3 start co-ordinate in the Full Sequence (V-genes only)")
    v_rs_start = IntegerField('v_rs_start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of V recombination site (V-genes only)")
    v_rs_end = IntegerField('v_rs_end', [validators.Optional()], description="End co-ordinate in the Full Sequence of V recombination site (V-genes only)")
    d_rs_3_prime_start = IntegerField('d_rs_3_prime_start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of 3 prime D recombination site (D-genes only)")
    d_rs_3_prime_end = IntegerField('d_rs_3_prime_end', [validators.Optional()], description="End co-ordinate in the Full Sequence of 3 prime D recombination site (D-genes only)")
    d_rs_5_prime_start = IntegerField('d_rs_5_prime_start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of 5 prime D recombination site (D-genes only)")
    d_rs_5_prime_end = IntegerField('d_rs_5_prime_end', [validators.Optional()], description="End co-ordinate in the Full Sequence of 5 prime D recombination site (D-genes only)")
    j_rs_start = IntegerField('j_rs_start', [validators.Optional()], description="Start co-ordinate in the Full Sequence of J recombination site (J-genes only)")
    j_rs_end = IntegerField('j_rs_end', [validators.Optional()], description="End co-ordinate in the Full Sequence of J recombination site (J-genes only)")
    j_codon_frame = SelectField('Codon Frame', choices=[('1', '1'), ('2', '2'), ('3', '3')], description="Codon position of the first sequence symbol in the Coding Sequence. Mandatory for J genes. Not used for V or D genes. ('1' means the sequence is in-frame, '2' means that the first bp is missing from the first codon, '3' means that the first 2 bp are missing)")
    j_cdr3_end = IntegerField('J CDR3 End', [validators.Optional()], description="In the case of a J-gene, the co-ordinate in the Coding Sequence of the first nucelotide of the conserved PHE or TRP (IMGT codon position 118)")
    c_exon_1_start = IntegerField('Exon 1 Start', [validators.Optional()], description="Start coordinate in the sequence field of the first constant region exon (C-genes only).")
    c_exon_1_end = IntegerField('Exon 1 End', [validators.Optional()], description="End coordinate in the sequence field of the first constant region exon (C-genes only).")
    c_exon_2_start = IntegerField('Exon 2 Start', [validators.Optional()], description="Start coordinate in the sequence field of the second constant region exon (C-genes only).")
    c_exon_2_end = IntegerField('Exon 2 End', [validators.Optional()], description="End coordinate in the sequence field of the second constant region exon (C-genes only).")
    c_exon_3_start = IntegerField('Exon 3 Start', [validators.Optional()], description="Start coordinate in the sequence field of the third constant region exon (C-genes only).")
    c_exon_3_end = IntegerField('Exon 3 End', [validators.Optional()], description="End coordinate in the sequence field of the third constant region exon (C-genes only).")
    c_exon_4_start = IntegerField('Exon 4 Start', [validators.Optional()], description="Start coordinate in the sequence field of the fourth constant region exon (C-genes only).")
    c_exon_4_end = IntegerField('Exon 4 End', [validators.Optional()], description="End coordinate in the sequence field of the fourth constant region exon (C-genes only).")
    c_exon_5_start = IntegerField('Exon 5 Start', [validators.Optional()], description="Start coordinate in the sequence field of the fifth constant region exon (C-genes only).")
    c_exon_5_end = IntegerField('Exon 5 End', [validators.Optional()], description="End coordinate in the sequence field of the fifth constant region exon (C-genes only).")
    c_exon_6_start = IntegerField('Exon 6 Start', [validators.Optional()], description="Start coordinate in the sequence field of the sixth constant region exon (C-genes only).")
    c_exon_6_end = IntegerField('Exon 6 End', [validators.Optional()], description="End coordinate in the sequence field of the sixth constant region exon (C-genes only).")
    c_exon_7_start = IntegerField('Exon 7 Start', [validators.Optional()], description="Start coordinate in the sequence field of the seventh constant region exon (C-genes only).")
    c_exon_7_end = IntegerField('Exon 7 End', [validators.Optional()], description="End coordinate in the sequence field of the seventh constant region exon (C-genes only).")
    c_exon_8_start = IntegerField('Exon 8 Start', [validators.Optional()], description="Start coordinate in the sequence field of the eighth constant region exon (C-genes only).")
    c_exon_8_end = IntegerField('Exon 8 End', [validators.Optional()], description="End coordinate in the sequence field of the eighth constant region exon (C-genes only).")
    utr_3_prime_start = IntegerField('UTR 3\' Start', [validators.Optional()], description="Start co-ordinate in the sequence field of the 3 prime UTR (V-genes only).")
    utr_3_prime_end = IntegerField('UTR 3\' End', [validators.Optional()], description="End co-ordinate in the sequence field of the 3 prime UTR (V-genes only).")
    paralogs = StringField('Paralogs', [validators.Length(max=255)], description="Canonical names of 0 or more paralogs")
    paralog_rep = BooleanField('Paralog Rep', [], description="If true, use the name of this allele in sequence-distinct datasets. If no representative is specified, the paralog with the lowest gene number is used.")
    inferred_extension = BooleanField('Extension?', [], description="Checked if there is an inferred extension to the canonical sequence")
    ext_3prime = TextAreaField('3\'  Extension', [ValidNucleotideSequence(ambiguous=False, gapped=True), validators.Optional()], description="Extending sequence at 3\' end (IMGT gapped)")
    start_3prime_ext = IntegerField('3\' start', [validators.Optional()], description="Start co-ordinate of 3\' extension (if any) in IMGT numbering")
    end_3prime_ext = IntegerField('3\' end', [validators.Optional()], description="End co-ordinate of 3\' extension (if any) in IMGT numbering")
    ext_5prime = TextAreaField('5\' Extension', [ValidNucleotideSequence(ambiguous=False, gapped=True), validators.Optional()], description="Extending sequence at 5\' end (IMGT gapped)")
    start_5prime_ext = IntegerField('5\' start', [validators.Optional()], description="Start co-ordinate of 5\' extension (if any) in IMGT numbering")
    end_5prime_ext = IntegerField('5\' end', [validators.Optional()], description="End co-ordinate of 5\' extension (if any) in IMGT numbering")
    curational_tags = SelectField('curational_tags', choices=[('none', 'none'), ('likely_truncated', 'likely_truncated'), ('likely_full-length', 'likely_full-length')], description="Controlled-vocabulary tags applied to this description")


